plugins/synaptics-rmi/fu-synaptics-rmi-v5-device.c

```diff
@@ -118,19 +140,164 @@ fu_synaptics_rmi_v5_device_write_block (FuSynapticsRmiDevice *self,
+			g_prefix_error (error, "failed to read status: ");
+			return FALSE;
+		}
+		if (!fu_synaptics_rmi_device_enter_backdoor (self, error))
+			return FALSE;
+		for (guint16 block_num = 0; block_num < rsa_block_cnt; block_num++) {
+			g_autoptr(GByteArray) res = NULL;
+			res = fu_synaptics_rmi_device_read_packet_register (self,
+									    f34->query_base + 14, /* addr of flash properties + 5 */
+									    0x3,
+									    error);
+			if (res == NULL)
+				return FALSE;
+			if (res->len != 0x3)
+				g_debug ("read %u bytes in return", res->len);
+			if (rsa_block_remain && block_num + 1 == rsa_block_cnt) {
+				g_byte_array_remove_range (res,
+							   rsa_block_remain,
+							   res->len - rsa_block_remain);
+			}
+			for (guint i = 0 ; i < res->len / 2 ; i++) {
+				guint8 tmp = res->data[i];
+				res->data[i] = res->data[res->len - i - 1];
+				res->data[res->len - i - 1] = tmp;
+			}
+			if (rsa_block_remain && block_num + 1 == rsa_block_cnt) {
+				g_byte_array_prepend (rsadump, res->data, rsa_block_remain);
+			} else {
+				g_byte_array_prepend (rsadump, res->data, res->len);
+			}
+		}
+		if (rsa_pubkey_len != rsadump->len) {
+			if (retries++ > 2) {
+				g_set_error (error,
+					     G_IO_ERROR, G_IO_ERROR_FAILED,
+					     "RSA public key length not matched %u: after %u retries: ",
+					     rsadump->len, retries);
+				return FALSE;
+			}
+			g_byte_array_set_size (rsadump, 0);
+			continue;
+		}
+
+		/* success */
+		break;
+	}
+	if (g_getenv ("FWUPD_SYNAPTICS_RMI_VERBOSE") != NULL) {
+		fu_common_dump_full (G_LOG_DOMAIN, "RSA public key",
+				     rsadump->data, rsadump->len,
+				     16, FU_DUMP_FLAGS_NONE);
+	}
+
+	/* sanity check size */
+	if (rsa_pubkey_len != rsadump->len) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "RSA public key length did not match: %u != %u: ",
+			     rsa_pubkey_len, rsadump->len);
+		return FALSE;
+	}
+
+	/* hash firmware data */
+	hash_data = gnutls_malloc (hash_length);
+	gnutls_hash_init (&sha2, GNUTLS_DIG_SHA256);
+	gnutls_hash (sha2, g_bytes_get_data (payload, NULL), g_bytes_get_size (payload));
+	gnutls_hash_deinit (sha2, hash_data);
+
+	/* hash */
+	hash.size = hash_length;
+	hash.data = hash_data;
+
+	/* modulus */
+	m.size = rsadump->len;
+	m.data = rsadump->data;
+
+	/* exponent */
+	e.size = sizeof(exponent);
+	e.data = exponent;
+
+	/* signature */
+	sig.size = g_bytes_get_size (signature);
+	sig.data = (guint8 *) g_bytes_get_data (signature, NULL);
+
+	gnutls_pubkey_init (&pub);
+	ec = gnutls_pubkey_import_rsa_raw (pub, &m, &e);
+	if (ec < 0) {
+		g_prefix_error (error, "failed to import RSA key: ");
+		return FALSE;
+	}
+	ec = gnutls_pubkey_verify_hash2 (pub, GNUTLS_SIGN_RSA_SHA256,
+					 0, &hash, &sig);
+	if (ec < 0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "failed to verify firmware");
+		return FALSE;
+	}
```

Comment:
I wonder if this is code that we actually want in the daemon for future use.  It seems pretty useful to me to use gnutls to verify hashes for binaries, and something we might want to do later.

Reply:
It solves the problem of the plugin needing gnutls... although I'm worried about hardcoding things like the exponent. Maybe this belongs in libjcat or something?

Reply:
exponent can be a function argument perhaps?

Reply:
@blueue could you write some gnutls example code to actually generate this signature please, and then at least we can write some unit tests for it before we work out where it needs to live. Thanks.

Reply:
@hughsie I found an API of gnutls that can use private key to sign hash data and get the related signature as well.
https://www.gnu.org/software/gnutls/reference/gnutls-abstract.html#gnutls-privkey-sign-data

Reply:
I looked into this a bit over the weekend; I don't want libjcat to basically wrap the API of gnutls_pubkey_verify_hash2 -- instead I've moved the gnutls verification to synaptics-rmi-common. If another plugin starts doing the same verification we can move it out to a libfwupdplugin shared function but until then I think it's fine where it is.